### DISCUSSION TITLE
How should we deal optimistic updates?

### DISCUSSION THREAD

**Discussion by chen-rn** [pmndrs/jotai#2114] [Q&A]
Hey all!

I'm fairly new to Jotai and is trying to wrap my head around how to deal with optimistic updates.

I have a mobile application that is fetching the user information when the user first opens the app, and storing that in an Atom. I'm doing that using the AtomWithQuery util!

Here's the code chunk

```
export const [, userQueryAtom] = atomsWithQuery(() => ({
  queryKey: [QUERY_KEYS.USER],
  queryFn: getCurrentUser,
}));
```

However, there are situations in which I'd like a client to instantaneously update part of that fetched user object.

I was hoping that I can use userQueryAtom like a normal atom like this
`const setUserQueryAtom = useSetAtom(userQueryAtom)`

Then set the atom
`setUserQuery((prev) => ({
        ...prev,
        info: {},//say if our goal was to reset this
      }));
`

But I realized that is not possible, since the atomsWithQuery are read ONLY atoms.

How should I approach something like this?

**Comment by chen-rn** [comment]
I realized that this might not be the responsibility of Jotai as we can do the same thing as if we're using React Query.

https://tanstack.com/query/v4/docs/react/guides/optimistic-updates

**Comment by tomschulze** [comment]
How would you implement the tanstack query [optimistic update example with rollbacks](https://tanstack.com/query/latest/docs/framework/react/guides/optimistic-updates#updating-a-list-of-todos-when-adding-a-new-todo) on error?

The callback signatures in `atomWithMutation` differ from the `mutationOptions`. While I could get the `queryClient` instance via the provided atom getter, I do not have access to the `onMutateResult` in the `onError` callback.
